File: src/main/java/az/millers/hcm/selfservice/domain/Announcement.java
package az.millers.hcm.selfservice.domain;

import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.UUID;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import lombok.Data;

@Entity
@Table(name = "announcement", schema = "selfservice")
@Data
public class Announcement {

    @Id
    @Column(name = "id")
    private UUID id;

    @Column(name = "tenant_id", nullable = false)
    private String tenantId = "default";

    @Column(name = "title", nullable = false, length = 300)
    private String title;

    @Column(name = "body", length = 4000)
    private String body;

    @Column(name = "publish_from", nullable = false)
    private LocalDate publishFrom;

    @Column(name = "publish_to")
    private LocalDate publishTo;

    @Enumerated(EnumType.STRING)
    @Column(name = "audience", nullable = false, length = 20)
    private AnnouncementAudience audience = AnnouncementAudience.ALL;

    @Column(name = "audience_ref")
    private UUID audienceRef;

    @Column(name = "active", nullable = false)
    private Boolean active = true;

    @Column(name = "created_at", nullable = false)
    private OffsetDateTime createdAt;

    @Column(name = "created_by", length = 80)
    private String createdBy;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @Column(name = "updated_by", length = 80)
    private String updatedBy;

    @PrePersist
    public void prePersist() {
        if (id == null) {
            id = UUID.randomUUID();
        }
        OffsetDateTime now = OffsetDateTime.now();
        if (createdAt == null) {
            createdAt = now;
        }
        if (updatedAt == null) {
            updatedAt = now;
        }
    }
}

File: mobile/lib/models/announcement.dart
/// Company announcement returned by GET /api/self/announcements (M430).
class Announcement {
  final String id;
  final String title;
  final String body;
  final String? publishFrom;
  final String? publishTo;
  final String? audience;
  final bool active;

  const Announcement({
    required this.id,
    required this.title,
    required this.body,
    this.publishFrom,
    this.publishTo,
    this.audience,
    this.active = true,
  });

  factory Announcement.fromJson(Map<String, dynamic> j) => Announcement(
        id: j['id'] as String,
        title: j['title'] as String? ?? '',
        body: j['body'] as String? ?? '',
        publishFrom: j['publishFrom'] as String?,
        publishTo: j['publishTo'] as String?,
        audience: j['audience'] as String?,
        active: j['active'] as bool? ?? true,
      );
}
